cirq/protocols/apply_unitary.py

```diff
@@ -85,28 +85,138 @@ def __init__(self,
+        """A default instance starting in state |0⟩.
+
+        Specify exactly one argument.
+
+        Args:
+            num_qubits: The number of qubits to make space for in the state.
+            qid_shape: The shape of the state, specifying the dimension of each
+                qid."""
+        if (num_qubits is None) == (qid_shape is None):
+            raise TypeError('Specify exactly one of num_qubits or qid_shape.')
+        if num_qubits is not None:
+            qid_shape = (2,) * num_qubits
+        qid_shape = cast(Tuple[int, ...], qid_shape)  # Satisfy mypy
+        num_qubits = len(qid_shape)
         state = linalg.one_hot(index=(0,) * num_qubits,
-                               shape=(2,) * num_qubits,
+                               shape=qid_shape,
                                dtype=np.complex128)
         return ApplyUnitaryArgs(state, np.empty_like(state), range(num_qubits))
 
-    def subspace_index(self, little_endian_bits_int: int
-                       ) -> Tuple[Union[slice, int, 'ellipsis'], ...]:
+    def for_operation_with_qid_shape(self, indices: Iterable[int],
+                                     qid_shape: Tuple[int, ...]
+                                    ) -> 'ApplyUnitaryArgs':
+        """Creates a sliced and transposed view of `self` appropriate for an
+        operation with shape `qid_shape` on qubits with the given indices.
+
+        Example:
+            sub_args = args.for_operation_with_qid_shape(indices, (2, 2, 2))
+            # Slice where the first qubit is |1>.
+            sub_args.target_tensor[..., 1, :, :]
+
+        Args:
+            indices: Integer indices into `self.axes` specifying which qubits
+                the operation applies to.
+            qid_shape: The qid shape of the operation, the expected number of
+                quantum levels in each qubit the operation applies to.
+
+        Returns: A new `ApplyUnitaryArgs` where `sub_args.target_tensor` and
+            `sub_args.available_buffer` are sliced and transposed views of
+            `self.target_tensor` and `self.available_buffer` respectively.
+        """
+        slices = [slice(0, size) for size in qid_shape]
+        sub_axes = [self.axes[i] for i in indices]
+        axis_set = set(sub_axes)
+        other_axes = [
+            axis for axis in range(len(self.target_tensor.shape))
+            if axis not in axis_set
+        ]
+        ordered_axes = (*other_axes, *sub_axes)
+        # Transpose sub_axes to the end of the shape and slice them
+        target_tensor = self.target_tensor.transpose(*ordered_axes)[(...,
+                                                                     *slices)]
+        available_buffer = self.available_buffer.transpose(*ordered_axes)[(
+            ..., *slices)]
+        new_axes = range(len(other_axes), len(ordered_axes))
+        return ApplyUnitaryArgs(target_tensor, available_buffer, new_axes)
+
+    def recover_result_from_sub_result(self,
```

Comment:
How about `incorporate_result`? We could then also include logic about swapping the target and the buffer, which is otherwise specified in a few places.

Does it make sense for this method to be on this type, exposed to users? We could move it into a private place.

Reply:
Renamed and made private and global.  This can't swap the target and the buffer because if you recursively use sub-arguments, swapping the target and the buffer messes up the second call to this.